packages/runtime/container-runtime/src/dataStoreContext.ts

```diff
@@ -280,11 +282,23 @@ export abstract class FluidDataStoreContext extends TypedEventEmitter<IFluidData
         this._attachState = this.containerRuntime.attachState !== AttachState.Detached && this.existing ?
             this.containerRuntime.attachState : AttachState.Detached;
 
+        /**
+         * The existing flag can be true in two conditions:
+         * 1. It's a local data store that is created when a detached container is rehydrated. In this case, the data
+         *    store is locally visible because the snapshot it is loaded from contains locally visible data stores only.
+         * 2. It's a remote data store that is created when an attached container is loaded is loaded from snapshot or
+         *    when an attach op comes in. In both these cases, the data store is already globally visible.
+         */
+        if (existing) {
+            this.visibilityState = this.containerRuntime.attachState === AttachState.Detached
```

Comment:
Is it worth adding validation function that can be called at various times and check some invariants? Like detached container can't have VisibilityState.GloballyVisible data store. Or bound/unbound lists in datastores class matches visibility of contexts. Maybe there are some other invariants that we can check, either locally (one data store / context) or globally (at root level for all nodes).

Reply:
I think it's useful (in addition to what you describe) to have validate() function that can be called any time and that validates that invariants are not broken.